lib/index.ts

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2016 Balena
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	 http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/**
+ * @module deviceStatus
+ */
+
+import keyBy = require('lodash/keyBy');
+import maxBy = require('lodash/maxBy');
+
+// This is the earliest possible year since
+// Balena didn't exist before that.
+const RESIN_CREATION_YEAR = 2013;
+
+/**
+ * @summary Map of possible device statuses
+ * @type {Object}
+ * @public
+ * @constant
+ */
+export const status = {
+	CONFIGURING: 'configuring',
+	IDLE: 'idle',
+	OFFLINE: 'offline',
+	INACTIVE: 'inactive',
+	POST_PROVISIONING: 'post-provisioning',
+	UPDATING: 'updating',
+} as const;
+
+export type StatusKey = typeof status[keyof typeof status];
+
+interface Status {
+	key: StatusKey;
+	name: string;
+}
+
+/**
+ * @summary Array of device statuses along with their display names
+ * @type {Object[]}
+ * @public
+ * @constant
+ */
+export const statuses: Status[] = [
+	// The order of statuses in this list is important, as it's reflected
+	// anywhere state are displayed -- currently in the device pie-chart.
+	{ key: status.IDLE, name: 'Online' },
+	{ key: status.CONFIGURING, name: 'Configuring' },
+	{ key: status.UPDATING, name: 'Updating' },
+	{ key: status.OFFLINE, name: 'Offline' },
+	{ key: status.POST_PROVISIONING, name: 'Post Provisioning' },
+	{ key: status.INACTIVE, name: 'Inactive' },
+];
+
+const statusesByKey = keyBy(statuses, 'key');
+
+export interface IHaveProgressAndStatus {
+	download_progress?: number;
+	status: string;
+}
+
+export interface Device extends IHaveProgressAndStatus {
+	is_active: boolean;
+	is_online: boolean;
+	last_connectivity_event?: string;
+	provisioning_progress?: number;
+	provisioning_state: string;
+	current_services?: {
+		[serviceName: string]: IHaveProgressAndStatus[];
+	};
+}
+
+/**
+ * @summary Get status of a device
+ * @function
+ * @public
+ *
+ * @param {Object} device - device
+ * @fulfil {Object} - device status
+ * @returns {Promise}
+ *
+ * @example
+ * balena = require('balena-sdk')
+ * deviceStatus = require('balena-device-status')
+ *
+ * balena.models.device.get('9174944').then (device) ->
+ * 	deviceStatus.getStatus(device).then (status) ->
+ * 		console.log(status.key)
+ * 		console.log(status.name)
+ */
+export const getStatus = (device: Device): Status => {
```

Comment:
Do we need the explicit return type?

Reply:
Actually we don't, but I would prefer to be sure that this has the exact intended type in the built typings.
Moreover, this way we will notice it early if by accident we add a return statement that alters the return type.